sklearn/preprocessing/_polynomial.py

```diff
@@ -250,55 +253,86 @@ def fit(self, X, y=None):
             "constant",
             "linear",
             "continue",
+            "periodic",
         ):
             raise ValueError(
                 "extrapolation must be one of 'error', "
-                "'constant', 'linear' or 'continue'."
+                "'constant', 'linear', 'continue' or 'periodic'."
             )
 
         if not isinstance(self.include_bias, (bool, np.bool_)):
             raise ValueError("include_bias must be bool.")
 
         # number of knots for base interval
         n_knots = base_knots.shape[0]
+
+        if self.extrapolation == "periodic" and n_knots <= self.degree:
+            raise ValueError(
+                "Periodic splines require degree < n_knots. Got n_knots="
+                f"{n_knots} and degree={self.degree}."
+            )
+
         # number of splines basis functions
-        n_splines = n_knots + self.degree - 1
+        if self.extrapolation != "periodic":
+            n_splines = n_knots + self.degree - 1
+        else:
+            # periodic splines have self.degree less degrees of freedom
+            n_splines = n_knots - 1
+
         degree = self.degree
         n_out = n_features * n_splines
         # We have to add degree number of knots below, and degree number knots
         # above the base knots in order to make the spline basis complete.
-        # Eilers & Marx in "Flexible smoothing with B-splines and  penalties"
-        # https://doi.org/10.1214/ss/1038425655 advice against repeating first
-        # and last knot several times, which would have inferior behaviour at
-        # boundaries if combined with a penalty (hence P-Spline). We follow
-        # this advice even if our splines are unpenalized.
-        # Meaning we do not:
-        # knots = np.r_[np.tile(base_knots.min(axis=0), reps=[degree, 1]),
-        #              base_knots,
-        #              np.tile(base_knots.max(axis=0), reps=[degree, 1])
-        #              ]
-        # Instead, we reuse the distance of the 2 fist/last knots.
-        dist_min = base_knots[1] - base_knots[0]
-        dist_max = base_knots[-1] - base_knots[-2]
-        knots = np.r_[
-            linspace(
-                base_knots[0] - degree * dist_min,
-                base_knots[0] - dist_min,
-                num=degree,
-            ),
-            base_knots,
-            linspace(
-                base_knots[-1] + dist_max,
-                base_knots[-1] + degree * dist_max,
-                num=degree,
-            ),
-        ]
+
+        if self.extrapolation == "periodic":
+            # For periodic splines the spacing of the first / last degree knots
+            # needs to a continuation of the base_knot spacing.
+            period = base_knots[-1] - base_knots[0]
+            knots = np.r_[
+                base_knots[-(degree + 1): -1] - period,
+                base_knots,
+                base_knots[1: (degree + 1)] + period
+            ].astype(np.float)
```

Comment:
This is causing the test failure. What is the dtype of `base_knots` at this point? It should already be `X.dtype` which should be fine, no? If you really need double precision, we can use `.astype(np.float64)` here instead but we should explain why. Usually scikit-learn models tends to operate at the precision level of the input data.

Reply:
If we pass knots as a list of lists of integers or an integer array, scipy complains:
```python
In [1]: from scipy.interpolate import BSpline
   ...: import numpy as np
   ...: 
   ...: BSpline.construct_fast(
   ...:     np.array([0, 1, 2, 3, 4, 5]),
   ...:     np.array([1.0, 0.0, 0.0]),
   ...:     2
   ...: )(np.array([1.0]))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-6975725b9102> in <module>
      2 import numpy as np
      3 
----> 4 BSpline.construct_fast(
      5     np.array([0, 1, 2, 3, 4, 5]),
      6     np.array([1.0, 0.0, 0.0]),

~/anaconda3/envs/sklearn/lib/python3.8/site-packages/scipy/interpolate/_bsplines.py in __call__(self, x, nu, extrapolate)
    336         out = np.empty((len(x), prod(self.c.shape[1:])), dtype=self.c.dtype)
    337         self._ensure_c_contiguous()
--> 338         self._evaluate(x, nu, extrapolate, out)
    339         out = out.reshape(x_shape + self.c.shape[1:])
    340         if self.axis != 0:

~/anaconda3/envs/sklearn/lib/python3.8/site-packages/scipy/interpolate/_bsplines.py in _evaluate(self, xp, nu, extrapolate, out)
    346 
    347     def _evaluate(self, xp, nu, extrapolate, out):
--> 348         _bspl.evaluate_spline(self.t, self.c.reshape(self.c.shape[0], -1),
    349                 self.k, xp, nu, extrapolate, out)
    350 

_bspl.pyx in scipy.interpolate._bspl.evaluate_spline()

ValueError: Buffer dtype mismatch, expected 'const double' but got 'long'
```

This hasn't been an issue before as `np.linspace` returns `floats` and `knots` get casted. [This test](https://github.com/scikit-learn/scikit-learn/pull/19483/commits/b8da79ff21e20fa029256b0a744bf57712f68404#diff-5527bc676d1f3e3d77d2964f863103039ef46f809db3a41ffaea58d5d686fb1bR90-R107) fails for periodic extrapolation. We could avoid the `.astype(np.float)` by using the default constructor of `BSpline`.

Reply:
`base_knots` are already `np.float64` as returned by `_get_base_knot_positions`. The reason is, again, that `BSpline` converts them anyway:
https://github.com/scipy/scipy/blob/5ab7426247900db9de856e790b8bea1bd71aec49/scipy/interpolate/_bsplines.py#L176

I think, the explicit conversion here is unnecessary.

Reply:
Not if the knots get passed explicitly (instead of being constructed with `_get_base_knot_positions`).

Reply:
Right. I passed manually, `check_array` only makes sure that `base_knots` is numeric.
I think, the simplest would be to pass `dtype=np.float64` to `check_array`.